### PR TITLE
Fast reduction

### PR DESCRIPTION
**Pull request by jonas-lj**
Fast modular reduction using Barrett Reduction. Is about 28% than BigInteger::mod.

### CODE REVIEW

File: core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/field/BigIntegerModulus.java [hunk 1]
```diff
@@ -29,7 +32,19 @@
...
+  
+  /**
+   * Compute <i>x</i> modulus this modulus <i>M</i> assuming that <i>0 &le; x < M<sup>2</sup></i>
+   * 
+   * @param x A non-negative integer smaller than the modulus squared.
+   * @return A BigInteger equal to x mod M.
+   */
+  BigInteger reduceModThis(BigInteger x) {
+    // We don't serialize the reducer and create it lazily if needed
+    reducer = Objects.requireNonNullElse(reducer, new ModularReducer(value));
```
Comment (n1v0lg) [inline core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/field/BigIntegerModulus.java]: This is not Java 8 compatible.

File: core/src/test/java/dk/alexandra/fresco/framework/util/TestModularReduction.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package dk.alexandra.fresco.framework.util;
+
+import static org.junit.Assert.assertEquals;
+
+import java.math.BigInteger;
+
+import org.junit.Test;
+
+public class TestModularReduction {
```
Comment (n1v0lg) [inline core/src/test/java/dk/alexandra/fresco/framework/util/TestModularReduction.java]: We should at a minimum test for:

* Each modulus we plan on using in SPDZ (i.e., those in ModulusFinder)
* Non-negative inputs smaller than the modulus
* Edge-cases: zero input, input equal to modulus

There are probably more edge cases, but I haven't looked through the implementation yet.

File: core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/field/BigIntegerModulus.java (hunk 1 above)
Comment (jonas-lj) [reply core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/field/BigIntegerModulus.java]: Fixed

File: core/src/test/java/dk/alexandra/fresco/framework/util/TestModularReduction.java (hunk 2 above)
Comment (jonas-lj) [reply core/src/test/java/dk/alexandra/fresco/framework/util/TestModularReduction.java]: I extended the tests (which caught an error :))

File: core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/field/BigIntegerModulus.java [hunk 3]
```diff
@@ -29,7 +32,21 @@
...
+  
+  /**
+   * Compute <i>x</i> modulus this modulus <i>M</i> assuming that <i>0 &le; x < M<sup>2</sup></i>
+   * 
+   * @param x A non-negative integer smaller than the modulus squared.
+   * @return A BigInteger equal to x mod M.
+   */
+  BigInteger reduceModThis(BigInteger x) {
+    // We don't serialize the reducer and create it lazily if needed
+    if (Objects.isNull(reducer)) {
```
Comment (pffrandsen) [inline core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/field/BigIntegerModulus.java]: why not just == null?

Comment (jonas-lj) [reply core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/field/BigIntegerModulus.java]: No reason. == null might be slightly more readable though, so I'll change it
